scalafix-core/shared/src/main/scala/scalafix/internal/rule/SingleAbstractMethod.scala

```diff
@@ -0,0 +1,174 @@
+package scalafix.internal.rule
+
+import scala.meta._
+import scalafix.Patch
+import scalafix.SemanticdbIndex
+import scalafix.rule.RuleCtx
+import scalafix.rule.SemanticRule
+
+/*
+http://www.scala-lang.org/files/archive/spec/2.12/06-expressions.html#sam-conversion
+
+SAM conversion
+
+An expression `(p1, ..., pN) => body` of function type `(T1, ..., TN) => T`
+  is sam-convertible to the expected type S if the following holds:
+
+ * the class C of S declares an abstract method m with signature (p1: A1, ..., pN: AN): R;
+ * besides m, C must not declare or inherit any other deferred value members;
+ * the method m must have a single argument list;
+ * there must be a type U that is a subtype of S, so that the expression
+  new U { final def m(p1: A1, ..., pN: AN): R = body } is well-typed
+  (conforming to the expected type S);
+ * for the purpose of scoping, m should be considered a static member
+  (U's members are not in scope in body);
+ * (A1, ..., AN) => R is a subtype of (T1, ..., TN) => T
+  (satisfying this condition drives type inference of unknown type parameters in S);
+
+Note that a function literal that targets a SAM is not necessarily compiled to the above
+instance creation expression. This is platform-dependent.
+
+It follows that:
+
+ * if class C defines a constructor, it must be accessible and must define exactly one, empty,
+  argument list;
+ * class C cannot be final or sealed
+  (for simplicity we ignore the possibility of SAM conversion in the same compilation
+   unit as the sealed class);
+ * m cannot be polymorphic;
+ * it must be possible to derive a fully-defined type U from S by inferring any unknown
+  type parameters of C.
+
+Finally, we impose some implementation restrictions (these may be lifted in future releases):
+
+ * C must not be nested or local (it must not capture its environment, as that results
+  in a zero-argument constructor)
+ * C's constructor must not have an implicit argument list (this simplifies type inference);
+ * C must not declare a self type (this simplifies type inference);
+ * C must not be @specialized.
+
+ */
+case class SingleAbstractMethod(index: SemanticdbIndex)
+    extends SemanticRule(index, "SingleAbstractMethod") {
+  override def description: String = ???
+  override def fix(ctx: RuleCtx): Patch = {
+
+    // extract the top name position java.util.TimerTask => TimerTask
+    def parentNamePosition(tpe: Type): Option[Position] = {
```

Comment:
This should already be handled by `index.symbol`. In my scope branch I've added more cases so `index.symbol` just works. Should maybe PR those immediately.

Reply:
Anything that is a subtype of `Ref` should be automatically handled by `index.symbol`, that includes `Init`

Reply:
Yep it look like it changed since I started this PR. `index.symbol(clazz)` now works